extension/js/common/oauth2/oauth2_finish.ts

```diff
@@ -0,0 +1,4 @@
+/* ©️ 2016 - present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com */
+import { OAuth2 } from './oauth2.js';
+
+OAuth2.finishAuth(window.location.href);
```

Comment:
I suppose this is where the `lastError` may happen, as showing in

```ts
  public static finishAuth = (url: string) => {
    const views = chrome.extension.getViews();
    // Loop through existing extension views and excute any stored callbacks.
    for (const view of views) {
      if ((view as any)['oauth-callback']) {
        (view as any)['oauth-callback'](url); // tslint:disable-line:no-unsafe-any
        (view as any)['oauth-login-finished'] = true;
      }
    }
    chrome.runtime.lastError = undefined;
    window.open('', '_self', '');
    window.close();
  };
```

It's a rather creative solution :-) but you don't have to do this. To communicate between views, you can use browser messages (`BrowserMsg` class). In this case, you can send a broadcast message `BrowserMsg.send.authWindowResult('broadcast', { any JSON serializable object you need })`. You'll need to add this `authWindowResult` to `BrowserMsg`, and then also add appropriate listeners in the pages that need this, instead of setting 

```ts
    (window.top as any)['oauth-login-finished'] = false;
    (window.top as any)['oauth-callback'] = callback;
```

That will make it more consistent with the rest of the codebase.

Now that you did it, I wonder if there is some advantage to what you've done compared to what we're doing, but that would be a discussion for another time.

Reply:
@rrrooommmaaa wdyt?

Reply:
> If we use `BrowserMsg.send` in `oauth2.ts` and add listener in `GoogleAuth` we won't benefit callback(promise) feature.

Why, can't the message handler resolve a promise?

Reply:
Let's talk about above code and let's say we use `BrowserMsg.send` to communicate between views. 
We might have `GoogleAuth.didGetOAuthResult` listener or something like that.
In `getOAuthResult` function, we launch oauth2 dialog by calling `GoogleAuth.oauthLogin` and we can't continue logic from here as we don't have success/failure callback.
Instead we have to add below logic into new listener method(GoogleAuth.didGetOAuthResult).

This way, google login logic will be seperate into `getOAuthResult` and `didGetOAuthResult` and it seems hard to read (at least for me)

```ts
if (!responseUrl) {
  return { acctEmail, result: 'Denied', error: 'Invalid response url', id_token: undefined };
}
const url = new URL(responseUrl!);
const allowedScopes = this.getParameterByName(url.search, 'scope');
const code = this.getParameterByName(url.search, 'code') ?? '';
if (!allowedScopes?.includes(this.OAUTH.scopes.compose) || !allowedScopes?.includes(this.OAUTH.scopes.modify)) {
...
```

Reply:
@ioanmo226 You're setting up a listener inside `launchWebAuthFlow`. Can you also set up a listener to a message that would be sent by `robots.txt`?

Reply:
@tomholub , @rrrooommmaaa I tried to migrate to `BrowserMsg.send` and here's a commit. (Below code is temporary, like using localStorage to save datas)
https://github.com/FlowCrypt/flowcrypt-browser/pull/4665/commits/b682f9af7338628724bce4c4068fd200c9f4b290

To me, updated code looks super messy and hard to understand.
I prefer previous way using callback method and want to revert change.

Let me know what you think.

Reply:
Will mark this as resolved as @rrrooommmaaa pushed fix